R/utils.R

```diff
@@ -11,132 +9,113 @@
+    # is the end of the last write or writeWS, unless eatWS() was called).
     readAll = function() {
-      private$writer$readAll()
-    },
-    # Removes both recent and upcoming whitespace writes
-    eatWS = function() {
-      # Undo recent whitespace writes
-      private$writer$restorePosition()
-      # Ignore upcoming whitespace writes
-      private$suppressing <- TRUE
-    }
-  )
-)
+      # Collapse everything in the buffer
+      s <- paste(buffer[seq_len(position)], collapse="")
 
-#' TextWriter class
-#'
-#' A class that manages the gradual concatenation of text. It
-#' provides two additional features (over a normal connection)
-#' that are important to us for tag writing:
-#'
-#' 1. Text is automatically converted to UTF-8.
-#'
-#' 2. Ability to bookmark (save) a write position, and jump back
-#'    to it (restore). Only one bookmark at a time is allowed.
-#'    Restoring a bookmark essentially truncates content that
-#'    was written after the bookmark was set.
-#'
-#' @noRd
-#' @importFrom R6 R6Class
-TextWriter <- R6Class("TextWriter",
-  private = list(
-    con = "ANY",
-    marked = numeric(1),
-    position = numeric(1)
-  ),
-  public = list(
-    initialize = function() {
-      # We use a file() here instead of textConnection() or paste/c to
-      # avoid the overhead of copying, which is huge for moderately
-      # large numbers of calls to .$write(). Generally when you want
-      # to incrementally build up a long string out of immutable ones,
-      # you want to use a mutable/growable string buffer of some kind;
-      # since R doesn't have something like that (that I know of),
-      # file() is the next best thing.
-      private$con <- file("", "w+b", encoding = "UTF-8")
-      private$marked <- 0
-      # position could be obtained by calling `seek(where=NA)`, but it's too expensive
-      private$position <- 0
-    },
-    close = function() {
-      close(private$con)
-    },
-    # Write content
-    #
-    # @param text Single element character vector
-    write = function(text) {
-      # The text that is written to this TextWriter will be converted to
-      # UTF-8 using enc2utf8. The rendered output will always be UTF-8
-      # encoded.
-      raw <- charToRaw(enc2utf8(text))
+      # Combine with anuthing we'd already accumulated
+      s <- paste0(accumulated, s)
 
-      # Update our current position
-      private$position <- private$position + length(raw)
-
-      # This is actually writing UTF-8 bytes, not chars
-      writeBin(raw, private$con)
-    },
-    # Return the contents of the TextWriter, as a single element
-    # character vector, from the beginning to the current writing
-    # position (normally this is the end of the connection, unless
-    # restorePosition() was called).
-    readAll = function() {
-      seek(private$con, where = 0, origin = "start", rw = "read")
-      s <- readChar(private$con, private$position, useBytes = TRUE)
       Encoding(s) <- "UTF-8"
```

Comment:
This shouldn't be necessary, since everything in `buffer` and `accumulated` should already be UTF-8, and `s` therefore should be as well. It was necessary when we were writing binary to a file and the reading it out, because the result of `readChar` would have been in the native encoding.